enarx-keep-sev-shim/src/hostcall.rs

```diff
@@ -0,0 +1,215 @@
+            }
+        }
+    }
+
+    /// Causes a `#VMEXIT` for the host to process the data in the shared memory
+    ///
+    /// Returns the contents of the shared memory reply status, the host might have
+    /// written.
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    #[inline(always)]
+    unsafe fn hostcall(&mut self) -> Result<[Register<usize>; 2], libc::c_int> {
+        let mut port = Port::<u16>::new(SYSCALL_TRIGGER_PORT);
+        port.write(1 as u16);
+        (*self.0).msg.rep.into()
+    }
+
+    /// Write `bytes` to a host file descriptor `fd`
+    ///
+    /// # Panics
+    ///
+    /// Panics, if `bytes.len()` exceeds `Block::buf_capacity()`
+    ///
+    /// # Safety
+    ///
+    /// The parameters returned can't be trusted.
+    pub unsafe fn write(&mut self, fd: usize, bytes: &[u8]) -> Result<libc::c_int, libc::c_int> {
+        let count = bytes.len();
+        assert!(count <= Block::buf_capacity());
+
+        (*self.0).msg.req.num = Register::from_raw(libc::SYS_write as usize);
+        (*self.0).msg.req.arg[0] = Register::from_raw(fd);
+        (*self.0).msg.req.arg[1] = ShimPhysAddr::from(
+            ShimVirtAddr::try_from(Address::from((*self.0).buf.as_mut_ptr()))
+                .map_err(|_| libc::EFAULT)?,
+        )
+        .into();
+        (*self.0).msg.req.arg[2] = Register::from_raw(count as _);
+        (*self.0).buf[..count].copy_from_slice(&bytes[..count]);
+        self.hostcall().map(|r| r[0].raw() as _)
+    }
+
+    /// Exit the shim with a `status` code
+    ///
+    /// # Panics
+    ///
+    /// Panics, if the shim resumes to run.
+    #[inline(always)]
+    pub fn exit_group(&mut self, status: u32) -> ! {
+        unsafe {
+            (*self.0).msg.req.num = Register::from_raw(libc::SYS_exit_group as _);
+            (*self.0).msg.req.arg[0] = Register::from_raw(status as _);
+
+            let _ = self.hostcall();
+
+            unreachable!()
+        }
+    }
+}
+
+impl Drop for HostCall {
+    fn drop(&mut self) {
+        // Put back in the virtual address in the global variable
+        let prev = unsafe { HOSTCALL_VIRT_ADDR.swap(self.0 as _, Ordering::Release) };
+        assert_eq!(prev, 0);
+    }
+}
+
+/// Write `bytes` to a host file descriptor `fd`
+#[inline(always)]
+pub fn shim_write(fd: HostFd, bytes: &[u8]) -> Result<libc::c_int, libc::c_int> {
+    let fd = usize::try_from(fd.as_raw_fd()).map_err(|_| libc::EBADF)?;
+    for c in bytes.chunks(Block::buf_capacity()) {
+        let chunk_len = c.len();
+        let mut to_write = chunk_len;
+
+        let mut host_call = HostCall::try_lock().ok_or(libc::EIO)?;
+
+        loop {
+            let written = unsafe {
+                host_call.write(
+                    fd,
+                    &c[(chunk_len.checked_sub(to_write).ok_or(libc::EFAULT)?)..],
+                )?
+            };
+            // be careful with `written` as it is untrusted
```

Comment:
BTW, I don't intend this as a blocker for this PR. I'm just thinking aloud. It would be nice to have some way to have validation of untrusted values enforced by the type system.

Reply:
@npmccallum agreed, let's think about that for the future.

Reply:
@haraldh Please file a debt bug then mark this as resolved.

Reply:
https://github.com/enarx/enarx/issues/630